src/cswinrt/strings/WinRT.cs

```diff
@@ -1063,110 +1072,96 @@ private EventRegistrationToken AddEventHandlerNoLock(T handler)
-            EventRegistrationToken token = GetPreferredToken(handler);
+            // Get a registration token, making sure that we haven't already used the value. This should be quite
+            // rare, but in the case it does happen, just keep trying until we find one that's unused. Note that
+            // this mutable part of the token is just 32 bit wide (the lower 32 bits). The upper 32 bits are fixed.
+            //
+            // Note that:
+            //   - If there is a handler assigned to the generated initial token value, it is not necessarily
+            //     this handler.
+            //   - If there is no handler assigned to the generated initial token value, the handler may still
+            //     be registered under a different token.
+            //
+            // Effectively the only reasonable thing to do with this value is to use it as a good starting point
+            // for generating a token for handler.
+            //
+            // We want to generate a token value that has the following properties:
+            //   1. Is quickly obtained from the handler instance (in this case, it doesn't depend on it at all).
+            //   2. Uses bits in the upper 32 bits of the 64 bit value, in order to avoid bugs where code
+            //      may assume the value is really just 32 bits.
+            //   3. Uuses bits in the bottom 32 bits of the 64 bit value, in order to ensure that code doesn't
+            //      take a dependency on them always being 0.
+            //
+            // The simple algorithm chosen here is to simply assign the upper 32 bits the metadata token of the
+            // event handler type, and the lower 32 bits to an incremental counter starting from some arbitrary
+            // constant. Using the metadata token for the upper 32 bits gives us at least a small chance of being
+            // able to identify a totally corrupted token if we ever come across one in a minidump or other scenario.
+            //
+            // We should feel free to change this algorithm as other requirements / optimizations become available.
+            // This implementation is sufficiently random that code cannot simply guess the value to take a dependency
+            // upon it. (Simply applying the hash-value algorithm directly won't work in the case of collisions,
+            // where we'll use a different token value).
+            int tokenLow32Bits = m_low32Bits++;
 
 #if NET6_0_OR_GREATER
             // When on .NET 6+, just iterate on TryAdd, which allows skipping the extra
             // lookup on the last iteration (as the handler is added rigth away instead).
-            while (!m_tokens.TryAdd(token, handler))
+            while (!m_tokens.TryAdd(tokenLow32Bits, handler))
             {
-                token = new EventRegistrationToken { Value = token.Value + 1 };
+                tokenLow32Bits++;
```

Comment:
One possible concern I see is: in case the token was somehow already used, incrementing it here would mean we could "race past" the value in `m_low32Bits`, meaning that all subsequent registrations would then also go through unnecessary extra lookups as the value of those lower 32 bits "catches up" to the one that ended up being used here. Now, I can't say I know exactly when/how such a scenario might happen though, and whether it's a real concern 🤔

@jkotas thoughts on this? And if this is an actual concern, should we perhaps simply increment `m_low32Bits` for each loop iteration, so we're sure that the value we leave it on should be the very next "free" value we expect?

Reply:
Yes, I think it is better to increment `m_low32Bits` so that the next registration does not need to go through the values that are in use again. (It is a corner case, unlikely to happen in practice.)

Reply:
Fixed in https://github.com/microsoft/CsWinRT/pull/1448/commits/d17885086c5686da370ecc6f66b9b552d8d1fcd7 🙂